ipaserver/install/cainstance.py

```diff
@@ -463,8 +463,6 @@ def configure_instance(self, host_name, dm_password, admin_password,
                         self.step("exposing CA instance on LDAP",
                                   self.__expose_ca_in_ldap)
 
-                    self.step("migrating certificate profiles to LDAP",
-                              migrate_profiles_to_ldap)
                     self.step("importing IPA certificate profiles",
```

Comment:
Is migrate_profiles_ldap() needed at all then? It's executed as part of the upgrade. Speeding that up would be nice too.

On my hardware I got a 40-second reduction in install time.

Reply:
I remember we had an issue that the full upgrade code didn't run at install time, only LDAP upgrades.

Reply:
The server upgrader is on my TODO list. For now I'm focusing on the installer. There are several upgrade steps that may no longer be necessary or that should be skipped if the step has been executed by another server in the cluster already.

I think the upgrade step is required when updating a server from an old Dogtag version or upgrading a cluster.

Reply:
@abbra w.r.t. profile migration, Dogtag itself imports LDAP profiles upon start up (when using the LDAPProfileSubsystem, which in FreeIPA it is).

Wondering about why this migration code is called in the installer, perhaps it was required at some point (during development) after the `LDAPProfileSubsystem` was implemented, but before the PKI installer learned the import behaviour.  I cannot remember the exact reasons, but this is my working theory.

Reply:
It may well be the reason, indeed.  The code comes from you ;)
```
commit b24fe0eb733c68af4042cdd78fca6f609efe843b
Author: Fraser Tweedale <ftweedal@redhat.com>
Date:   Mon May 11 21:17:48 2015 -0400
```

Reply:
The commit where it was added is 620036d26e98fdcefff00168e9e5463a8257d49c.

Profiles are not being imported during *replica installation*, apart from this step.  So I think we still need it but only for replica installation, and we can skip it if we see that profiles are already in LDAP.

Reply:
> @abbra w.r.t. profile migration, Dogtag itself imports LDAP profiles upon start up (when using the LDAPProfileSubsystem, which in FreeIPA it is).

Could you please clarify. Does Dogtag import missing profiles on every startup of the pki-tomcat service or does Dogtag only import profiles during pkispawn?

Reply:
@tiran only during pkispawn, and *not* for cloning (replica install).

Reply:
@tiran stand by, I want to confirm that.

Reply:
The code block with ``migrate_profiles_to_ldap`` is guarded by ``if not promote``. Replica installer calls the function with ``promote=True``.

Reply:
@tiran confirmed.  From `base/server/python/pki/server/deployment/scriptlets/configuration.py`~912:
```python
        if not clone and subsystem.type == 'CA':
            subsystem.import_profiles( 
                input_folder='/usr/share/pki/ca/profiles/ca') 
```

(Sorry that took a while to work out.  The import mechanism has completely changed since I wrote it 6 years ago).

Reply:
@tiran wrote:
> The code block with migrate_profiles_to_ldap is guarded by if not promote. Replica installer calls the function with promote=True.

Let's think this through.

We no longer support DL0 replica installation right?  That means that the peer server is at a version that has LDAP profiles.  Therefore replica installation never needs to import profiles.  Dogtag CA cloning procedure does not import profiles to LDAP, but they are already in LDAP so that is not a problem.

Does the logic work out?  If it does, then we can indeed delete the profile import step for all installation scenarios.

Reply:
I confirmed removal of DL0 support with Martin. IPA 4.8 / RHEL 8 no longer support DL0 IdM clusters. DL0 was deprecated in [RHEL 7.9](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/8.0_release_notes/index). We can safely assume that a 4.9 replica requires a DL1 cluster that has been migrated to have LDAP profiles.

Reply:
OK :+1:

*hands back the ack*.